content/blog/2025-04-24-super-optimization/index.md

```diff
@@ -0,0 +1,104 @@
+
+<div style="display: flex; justify-content: center; gap: 10px;">
+  <img src="egraph-a.png" alt="" style="width: 40%;">
+  <img src="egraph-d.png" alt="" style="width: 40%;">
+</div>
+
+The e-graph is then converted to a system of constraints using an architecture specific timing model that encodes the cost of each operation. These constraints request a SAT solver to either prove “No program computes the result in k cycles”, or give a counterexample program. Denali uses binary search to find the least value of k that works.  The paper reports results from applying Denali to the problem of swapping the 4 lower bytes of a register, for which it took 1 minute to identify an optimal 5-cycle solution, as well as to a more complicated program computing a checksum which required 4 hours to solve.
+
+## Is the cost prohibitive?
+
+Clearly, Denali is limited by its relatively high running time. As illustrated by their checksum example, with a time budget of minutes to hours, they were only able to output relatively small programs, on the order of a couple dozen instructions. While both computers and tools have gotten faster over time, the problem is still fundamentally exponential: there will always be a point past which the superoptimization approach is infeasible.
+
+However, there are still situations when pouring a lot of resources into small pieces of code can be very useful, and techniques from superoptimization can influence / inspire normal compilers as well. This [overview of recent e-graph applications](https://github.com/philzook58/awesome-egraphs?tab=readme-ov-file#program-optimization) shows a bunch of recent (mainly academic) projects using e-graphs for program optimization, in varied domains, such as compiling tensor graphs, SQL queries, and GPU kernels.
+
+Not all of these are superoptimization: some just aim to apply similar techniques, but with the same goals as normal program optimization. In addition, algorithms originally developed for superoptimization have [been adapted](https://cfallin.org/pubs/egraphs2023_aegraphs_slides.pdf) for parts of the Cranelift compiler backend. On this note, it’s interesting to consider how superoptimization can be of use for normal compilers, more broadly. The paper “Automatic Generation of Peephole Superoptimizers” explored ways in which superoptimized general-purpose code snippets can be used to generate a database of transformations, and showed that it can enable signficant speed-ups for compiled code.
```

Comment:
Sorry for the repetition, but [the Bansal and Aiken paper](https://theory.stanford.edu/~aiken/publications/papers/asplos06.pdf) probably also deserves a link.